### ISSUE TITLE
Add Anchor Button to CKEditor

### ISSUE THREAD

**Issue by mikesoltis** [backdrop/backdrop-issues#4914]
**Description of the need**

I have many pages of instruction manuals for x-ray diffraction facilities that will incorporate anchors (from menus).  

**Proposed solution**

Add CKEditor Anchor button to toolbar selection.

**Alternatives that have been considered**

Manual input in source

**Additional information**

#1231 is complete, but it addressed removing the Anchor button that did not work.  I do not see it in the "to do" list.

**Draft of feature description for Press Release (1 paragraph at most)**

Backdrop now includes the CKEditor Anchor button.

**Comment by indigoxela** [comment]
@mikesoltis many thanks for taking the time to report this issue.

There seems to be a contrib module that ships with this button: [ckeditor_addon](https://github.com/backdrop-contrib/ckeditor_addon) (has no stable release, though).

> Alternatives that have been considered ... Manual input in source

I'd agree that this is no real alternative. Most users wouldn't be able to do it.

So, should we add it in core or is this rather a contrib candidate? More feedback is welcome.

**Comment by klonos** [comment]
> There seems to be a contrib module that ships with this button: [ckeditor_addon](https://github.com/backdrop-contrib/ckeditor_addon) (has no stable release, though, and I didn't test it).

I've tested it, and the anchor button does not get added 😞 (expected, as the module has not been updated since 2016) ...moreover, it downgrades CKEditor to v4.5.7, which is not secure, and could lead to regressions.

**Comment by olafgrabienski** [comment]
What is the anchor button supposed to do?

I'm asking because as of https://github.com/backdrop/backdrop-issues/issues/1777, the CKEditor link dialog in Backdrop allows linking to an ID of the same page.

**Comment by indigoxela** [comment]
> What is the anchor button supposed to do?

It adds a link with an ID to act as a target for on-page links like `#the-id-to-jump-to`

Note: adding an ID to a link *is* possible with the ckeditor link dialog from core, but it's not obvious and the link url validation further complicates things. Room for improvement. :wink:

**Comment by mikesoltis** [comment]
Features of the Anchor button:

Button opens dialog to insert title, saves at cursor position/selection.
Works with white spaces in anchor name.
Link module provides a tab to select from list of anchors on page.

**Comment by jenlampton** [comment]
I'd also like to see this in core. I have had almost every client ask me how they can add anchors, and some have resorted to switching to HTML mode to type in `<a name="">` not knowing that the name attribute is now obsolete for `a` tags. If we could provide a button that offered the same functionality, they would be doing it "the right way" without needing to understand the nuancnces of HTML5.

**Comment by klonos** [comment]
Some initial research: https://stackoverflow.com/questions/46434435/ckeditor-how-to-insert-anchor-tag-into-document

> Empty Links are removed from CKEDITOR automatically. You can add `data-cke-survive="true"` so these links aren't removed, `<a href="#" data-cke-survive="true"></a>`

**Comment by klonos** [comment]
...I was under the impression that we can use any element with an ID for anchor. Does it have to be an `<a>` tag?

**Comment by jenlampton** [comment]
Any element.

**Comment by mikesoltis** [comment]
There is an implementation for Drupal 8.  Can it be used to expedite porting to Backdrop?

https://github.com/Vardot/anchor_link

**Comment by yorkshire-pudding** [comment]
I agree that we need an easy way to make any element (I guess it will usually be headings but could be others) into an anchor destination by adding the `id` attribute.  I have to do this by switching to source but there is no way you can expect users to do that.  The D8 module seems to use the `name` attribute which as @jenlampton points out is now obsolete

**Comment by herbdool** [comment]
@yorkshire-pudding @olafgrabienski My PR enables the CKEditor 5 Bookmark button. https://ckeditor.com/docs/ckeditor5/latest/features/bookmarks.html (despite the docs saying it's a premium feature, it's already included in Backdrop, just not enabled).

**Comment by olafgrabienski** [comment]
@herbdool Thanks for the PR! I've tested the Bookmark button in the sandbox, and it works for me.

(Apart from that, I have no idea if we can use the premium feature question.)

**Example output:**

```
<h3>
    <a id="lorem-ipsum-dolor"></a>Lorem ipsum dolor
</h3>
```

**How it looks in CKEditor:**

<img width="250" src="https://github.com/user-attachments/assets/acdc23b3-6639-4c60-9bc3-e9f0cd4d185f" />

**Comment by herbdool** [comment]
Thanks @olafgrabienski. Can you update the labels if it works for you?

It's confusing that CKEditor still lists Bookmark as a "premium feature". Maybe it's outdated or maybe the integration with their Link dialog is the premium part. Regardless, it's already included in Backdrop and in Drupal (https://www.drupal.org/project/ckeditor5_bookmark) and it can be enabled without it saying the feature needs to be unlocked. So shrug?

**Comment by olafgrabienski** [comment]
Sure, updated the labels. The PR doesn't implement everything from https://github.com/backdrop/backdrop-issues/issues/4914#issuecomment-775676766, but the most important feature works fine:

> Button opens dialog to insert title, saves at cursor position/selection.

The other features of the Anchor button could be discussed in a follow-up:

> Works with white spaces in anchor name.
> Link module provides a tab to select from list of anchors on page.

**Comment by herbdool** [comment]
@olafgrabienski 

> Works with white spaces in anchor name.

We don't have much control over that if we use the built-in version. It does stop someone from entering white space, which is key since `id` attributes can't have spaces:

<img width="492" height="175" alt="Image" src="https://github.com/user-attachments/assets/9daa22d8-9a1f-4886-8d27-e7d7ce645b58" />

> Link module provides a tab to select from list of anchors on page.

I agree that should be a follow-up since the Link module has its own dialogue and doesn't use CKE for that.

**Comment by quicksketch** [comment]
We reviewed this today during the [weekly development meeting](https://www.youtube.com/watch?v=sfYM3BABxaQ). Tested and it works great! I love that this feature finally exists out-of-the-box!

There might be different ways that people might want to style the front-end or CKEditor behavior, but going with the default, minimal behavior seems like is the way to go (or at least where to start).

> (despite the docs saying it's a premium feature, it's already included in Backdrop, just not enabled).

It's been a while since I checked in on CKEditor's licensing/pricing. But wow, costs on plans from https://ckeditor.com/pricing/ seem outrageous to me. With pricing be "X editor loads per month", that would eliminate the ability for CKEditor to be used in a forum or any place where the editor is used on the front-end of a website (when using the commercial license). However, we don't use the commercial license. CKEditor is still also [licensed under the GPL, per their LICENSE.md file](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5/LICENSE.md). I don't really understand how something could be _both_ GPL and commercial, but it seems pretty clear that we're not beholden to the commercial restrictions when using it in an open-source context like this.

**Comment by avpaderno** [comment]
Premium features should not work for sites that use `'GPL'` as `licenseKey`. At least, I imagine that value tells them when a site uses Premium CKEditor.  
Even on https://ckeditor.com/pricing/#compare-features, Bookmarks is shown as "not available" for free. I am wondering if they do not mean "not available on the cloud-hosted solution" but I do not think that would make sense, if not using the cloud-hosted solution would be sufficient to use premium features.

**Comment by quicksketch** [comment]
Sorry the red X's and green checkmarks were not loading for me earlier, so I couldn't tell the difference. Now it's loading correctly.

<img width="1705" height="960" alt="Image" src="https://github.com/user-attachments/assets/96431051-e867-4dda-9a5b-13b9338ad6d5" />

I'm almost certain we're already getting some of these other benefits like List Properties, Multi-level Lists, and Remove Formatting. Image Resize is also marked as a premium-only feature. I don't think this pricing page applies to GPL usage since it would contradict what has been available for years for sites that use the "GPL" license key.

I'll add that the Bookmark Plugin itself also has a [specific LICENSE.md that also notes it is GPL](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-bookmark/LICENSE.md). As it would need to anyway since everything that loads libraries that are GPL themselves become GPL. 

If CKEditor starts removing "premium" features from the GPL-licensed version, we'd loose a lot more than Bookmarks. I don't think this is an immediate problem, but I am glad that we have the TinyMCE contrib project waiting in the wings.

**Comment by indigoxela** [comment]
The confusion seems to be: self-hosted vs. using their cloud service, I guess.

We include the library, not using their service. They don't mention self-hosting on that page (OK, because they don't make money with it, but that's also true with Drupal integration).

> If CKEditor starts removing "premium" features from the GPL-licensed version, we'd loose a lot more than Bookmarks. 

For sure. But we have a plan B. 😉

**Comment by avpaderno** [comment]
Probably, the features available for a project that can use the GPL license for CKEditor are equivalent to the ones listed under *custom* on https://ckeditor.com/pricing/#compare-features. I assume it should be so because that is the only plan which allows self-hosting.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/5313 into 1.x for 1.35.0. Thank you @herbdool for the PR! Thank you @indigoxela, @avpaderno, @klonos, @mikesoltis, @jenlampton, @olafgrabienski, and @yorkshire-pudding for the discussion, review, and feedback!
